File: easy/1175/prime_arrangements.cpp

#define MAX_NUM 1000000007

class Solution {
public:
    long fact(int n) {
        long f = 1;
        
        if (n > 1) {
            for (int i = 2; i <= n; i++) {
                f = (f * i) % MAX_NUM;   
            }
        }
        
        return f % MAX_NUM;
    }

    int numPrimeArrangements(int n) {
        int not_prime = 1;
        vector<bool> num_list(n+1);

        for (int i = n; i > 1; i--) {
            for (int j = 2; j < n; j++) {
                if (i*j <= n) {
                    not_prime += num_list[i*j] ? 0 : 1;
                    num_list[i*j] = true;
                } else {
                    j = n;   
                }
            }
        }

        return (fact(not_prime) * fact(n - not_prime)) % MAX_NUM;
    }
};

File: easy/1175/prime_arrangements.py

class Solution:
    def fact(self, n: int) -> int:
        f = 1

        for i in range(2, n + 1):
            f *= i
                
        return f

    def numPrimeArrangements(self, n: int) -> int:
        not_prime = 1
        num_list = [False] * (n+1)

        for i in range(n, 1, -1):
            for j in range(2, n):
                if i*j <= n:
                    not_prime += 0 if num_list[i*j] else 1
                    num_list[i*j] = True
                else:
                    break

        return (self.fact(not_prime) * self.fact(n - not_prime)) % (10**9 + 7)

